storage: don't modify the given cfg.Opts

RaduBerinde opened the pull request:
This change improves the `NewPebble` code to not modify the given `cfg.Opts`. Such behavior is surprising and can trip up tests that reuse the same config.

Also, `ResolveEncryptedEnvOptions` and `wrapFilesystemMiddleware` no longer modify the `Options` directly; and `CheckNoRegistryFile` is now a standalone function.

Release note: None
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96770)
<!-- Reviewable:end -->

RaduBerinde committed:
storage: don't modify the given cfg.Opts

This change improves the `NewPebble` code to not modify the given
`cfg.Opts`. Such behavior is surprising and can trip up tests that
reuse the same config.

Also, `ResolveEncryptedEnvOptions` and `wrapFilesystemMiddleware` no
longer modify the Options directly; and `CheckNoRegistryFile` is now a
standalone function.

Release note: None
Epic: none

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

thanks for tackling this much-needed cleanup!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96770#-:-NNlXOs48Iv7lLxEwOKq:b-pggauk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[`pkg/storage/pebble.go` line 862 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96770#-NNlWpq57vB76Ppiyqqy:-NNlWpq57vB76Ppiyqqz:b-e3qynk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ee61fb97e47309617b0e626ac7ff3e1665eb30c/pkg/storage/pebble.go#L862)):*
> ```Go
> func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
> 	// pebble.Open also calls EnsureDefaults, but only after doing a clone. Call
> 	// EnsureDefaults beforehand so we have a matching cfg here for when we save
> ```

lol at how far this comment had drifted from the `cfg.Opts.EnsureDefaults()` call that it describes.


<!-- Sent from Reviewable.io -->
